tsinfer/inference.py

```diff
@@ -1499,105 +1477,86 @@ def __start_epoch(self, epoch_index):
-            self.__start_epoch(j)
-            start, end = map(int, self.epoch_slices[j])
-            for ancestor_id in range(start, end):
-                a = next(self.ancestors)
-                assert ancestor_id == a.id
-                self.__ancestor_find_path(a)
-            self.__complete_epoch(j)
-
-    def __match_ancestors_multi_threaded(self, start_epoch=1):
-        # See note on match samples multithreaded below. Should combine these
-        # into a single function. Possibly when trying to make the thread
-        # error handling more robust.
-        queue_depth = 8 * self.num_threads  # Seems like a reasonable limit
-        match_queue = queue.Queue(queue_depth)
 
-        def match_worker(thread_index):
-            while True:
-                work = match_queue.get()
-                if work is None:
-                    break
-                self.__ancestor_find_path(work, thread_index)
-                match_queue.task_done()
-            match_queue.task_done()
+    def match_ancestors(self):
+        logger.info(f"Starting ancestor matching for {self.num_epochs} epochs")
+        self.match_progress = self.progress_monitor.get("ma_match", self.num_ancestors)
 
-        match_threads = [
-            threads.queue_consumer_thread(
-                match_worker, match_queue, name=f"match-worker-{j}", index=j
+        local_data = threading.local()
+
+        def worker_function(ancestor_ancestor_id):
+            ancestor, ancestor_id = ancestor_ancestor_id
+            assert ancestor.id == ancestor_id
+            if not hasattr(local_data, "matcher"):
+                local_data.matcher = self.create_matcher_instance()
+            result = self.find_path(
+                matcher=local_data.matcher,
+                child_id=ancestor.id,
+                haplotype=ancestor.full_haplotype,
+                start=ancestor.start,
+                end=ancestor.end,
             )
-            for j in range(self.num_threads)
-        ]
-        logger.debug(f"Started {self.num_threads} match worker threads")
+            self.match_progress.update()
+            return result
 
         for j in range(self.start_epoch, self.num_epochs):
             self.__start_epoch(j)
             start, end = map(int, self.epoch_slices[j])
-            for ancestor_id in range(start, end):
-                a = next(self.ancestors)
-                assert a.id == ancestor_id
-                match_queue.put(a)
-            # Block until all matches have completed.
-            match_queue.join()
-            self.__complete_epoch(j)
+            if self.num_threads > 0:
+                mapper = lambda func, args: threads.threaded_map(  # noqa E731
+                    func, args, self.num_threads
+                )
+            else:
+                mapper = map
+            # We have to consume the iterator here, otherwise we'll match against
+            # ancestors in this epoch
+            results = list(
```

Comment:
Any reason not to do something like this?
```python
results = []
work_iter = zip(self.ancestors, range(start, end))
for result in mapper(worker_function, work_iter):
     results.append(result)
     match_progresss.update()
```

I always find it difficult to untangle nested iterators like that, and it will surely be helpful to take the progress update out of worker threads

Reply:
I had a go in https://github.com/tskit-dev/tsinfer/pull/829/commits/03b61b67716bab558054a1ad00c0aa198a44a93c